### PR TITLE
Add keybinding to rename tab

### LINKED ISSUE #6256: Add keybinding to edit tab name
Hi guys, I'm really excited to start using this feature!

I couldn't really find anything related to it, so I am asking here. Would it be possible to bind a shortcut key to the rename action, so that it open up the rename box once pressed?

I quite often rename my windows once I repurposed them and would love not to have to take my hands of the keyboard. Let me know if I should create an issue for it.

_Originally posted by @BartXZX in https://github.com/microsoft/terminal/pull/5775#issuecomment-636090446_

### PR DESCRIPTION
**Pull request by ggadget6**
## Summary of the Pull Request
Add keybinding for renaming a tab
 
## PR Checklist
* [X] Fulfills format requirements set by #6567 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [X] Tests passed
* [X] Requires documentation to be updated
* [X] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #6567 and here (#6557)

This no longer c loses #6256, as the spec changed.

### CODE REVIEW

**Comment by ggadget6** [comment]
Made it a draft because I'm currently working on it. Was wondering if I could get any more guidance on where to go--I started implementing some things related to ActionArgs but I'm a little confused on where to go from here, and about whether what I did was correct

**Comment by zadjii-msft** [comment]
You're definitely on the right path here. I think something important we'll need to decide how to handle the following scenarios:
1. how does the user set the tab text to a specific string?
2. how does the user clear the tab text (reverting it to the application's title)
3. how does the user jump to the tab rename box? (see #5775)

I'm doing something _just like this_ in #6567, only with the tab color instead of the title. After sending that PR, I think the right way to do with should be:
* `"command": { "action": "setTabTitle", "title":"someString" }`: set the tab title to "someString" (this is scenario 1)
* `"command": { "action": "setTabTitle", "title":"" }`: set the tab title to "" (this is _also_ scenario 1, just a unique edge case)
* `"command": { "action": "setTabTitle", "title":null }`: reset the tab title to whatever the application set. (this is scenario 2)
* `"command": { "action": "promptForTabTitle" }` (this action name would need to be better): Activate the tab rename text box (this is scenario 3)

Granted, the discussion is still pending in #6567. I'd want to make sure that both this and that feature settle on a singular consistent design.

I wouldn't mind if this PR only covered situation 1&2, and then a follow-up PR was made for scenario 3. I just want to make sure we consider the whole picture before committing to a single piece.

As far as actually implementing this bit - Providing public methods on `Tab` for `SetTabText`, `ResetTabText`, `ActivateTabRenameBox` (boy that name needs work) that set/reset `Tab::_runtimeTabText` should be enough to make this work

**Comment by ggadget6** [comment]
Ok, I worked on it for a while and got scenario 1 to work, except for in the edge case. I think I can get the edge case to work (I think I just need to use _tabViewItem.Header correctly, right?), but I'm currently not sure how to detect scenario 2. How can I detect that the user has input null?

**Comment by zadjii-msft** [comment]
@DHowett can correct me if I'm wrong, but you might be able to get that for free with:

`GETSET_PROPERTY(winrt::hstring, Title, nullptr);`

**Comment by ggadget6** [comment]
Yeah, looking closer I agree. The current behavior is what you talked about in your comment, so I'll mark this as ready for review.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 1]
```diff
@@ -1095,6 +1097,18 @@ namespace winrt::TerminalApp::implementation
         return std::nullopt;
     }
 
+
+    // Method Description:
+    // - returns a com_ptr to the currently focused tab. This might return null,
+    //   so make sure to check the result!
+    winrt::com_ptr<Tab> TerminalPage::_GetFocusedTab()
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1104]: Heads up - this'll probably merge conflict, since I added the same function in #6567 😋

File: src/cascadia/TerminalApp/AppActionHandlers.cpp [hunk 2]
```diff
@@ -235,4 +235,34 @@ namespace winrt::TerminalApp::implementation
...
     }
+
+    void TerminalPage::_HandleRenameTab(const IInspectable& /*sender*/,
+                                          const TerminalApp::ActionEventArgs& args)
+    {
+        std::optional<winrt::hstring> title;
+
+        if (const auto& realArgs = args.ActionArgs().try_as<TerminalApp::RenameTabArgs>())
+        {
+            //if (realArgs.Title() != nullptr)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/AppActionHandlers.cpp:246]: Get rid of the dead code here in this function

File: src/cascadia/TerminalApp/ActionArgs.h [hunk 3]
```diff
@@ -381,6 +382,35 @@ namespace winrt::TerminalApp::implementation
...
+        };
+        static FromJsonResult FromJson(const Json::Value& json)
+        {
+            // LOAD BEARING: Not using make_self here _will_ break you in the future!
+            auto args = winrt::make_self<RenameTabArgs>();
+            if (auto title{ json[JsonKey(TitleKey)] })
+            {
+               args->_Title = winrt::to_hstring(title.asString());
+            }
+            //args->_Title = RenameTabArgs::FromJson(json);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/ActionArgs.h:410]:
```suggestion
```

**Comment by ggadget6** [comment]
We had one big merge conflict in AppKeyBindingsSerialization.cpp because of #6351 . We resolved it, but our issue now is that the three lines we added to AppKeyBindingsSerialization.cpp weren't added to the new file, ActionAndArgs.cpp.  How should we deal with this?

**Comment by DHowett** [comment]
Aaah, it looks like your newly-merged files have line endings in the wrong (?) (or at least, different-from-what-we-have-right-now) format -- this might make reviewing and merging this difficult.

This might be an instance where you'll just have to update ActionAndArgs youself to re-add those lines (based on your earlier commit before the merge). It's easy to lose things during a merge conflict ☹️

**Comment by ggadget6** [comment]
It's weird that the line endings were wrong, I thought they were ok before I pushed them. I used GitHub's Merge conflict resolution tool to fix the conflict, is that possibly part of the issue? 

To fix the ActionAndArgs thing, would I just go back to the commit before the merge, modify ActionAndArgs, and git push --force onto my fork's branch? If I'm correct, I think this would let me fix the merge conflict using GitHub's merge conflict tool.

**Comment by DHowett** [comment]
GitHub's merge conflict resolution tool will, unfortunately, almost always mangle the files in this repository. ☹️ 

You might have better luck checking out the files locally and resolving the merge conflict locally.

**Comment by ggadget6** [comment]
Well, I tried to fix the line ending issue, but it seems like the CI test still doesn't like it... Is there any way to find out what file it's complaining about?

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 1 above)
Comment (ggadget6) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1103]: It didn't conflict, maybe because I copied it straight from #6567 into the exact place that you had put it.

File: src/cascadia/TerminalApp/ActionArgs.idl [hunk 4]
```diff
@@ -97,9 +97,14 @@ namespace TerminalApp
         NewTerminalArgs TerminalArgs { get; };
         SplitType SplitMode { get; };
     };
-
+
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/ActionArgs.idl:100]: nit: undo this change

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 5]
```diff
@@ -1096,6 +1097,17 @@ namespace winrt::TerminalApp::implementation
...
+    {
+        if (auto index{ _GetFocusedTabIndex() })
+        {
+            return _GetStrongTabImpl(*index);
+        }
+        return nullptr;
+    }
     // Method Description:
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1111]: nit: missing newline

**Comment by carlos-zamora** [comment]
Would you mind also adding a PR in [the docs repo](https://github.com/MicrosoftDocs/terminal) please (or at the very least an issue to track it)?

**Comment by ggadget6** [comment]
Just added an [issue](https://github.com/MicrosoftDocs/terminal/issues/74) in the docs repo, I'll make a PR soon when I get the chance!
